src/main/java/io/cdap/plugin/spark/HierarchyToRelationalUtils.java

```diff
@@ -0,0 +1,213 @@
+   * @return dataset with rows matching root record criteria
+   */
+  private Dataset<Row> generateTreeRoot(Dataset<Row> data, HierarchyToRelationalConfig config) {
+    String additionalFieldsForRoot = String.format("0 as %s,'%s' as %s,'%s' as %s ",
+                                                   config.levelField, config.trueValueField, config.topField,
+                                                   config.falseValueField, config.bottomField);
+    String parentProductField = "ParentProduct";
+    String childProductField = "ChildProduct";
+    String otherFields = "null as Supplier, null as Sales";
+    String selfAsParentQuery = SQLUtils.sqlSelfParentQuery(config, additionalFieldsForRoot, parentProductField,
+                                                           childProductField, otherFields,
+                                                           SQLUtils.NAME_HIERARCHY_TABLE);
+    return data.sqlContext().sql(selfAsParentQuery).limit(1);
+  }
+
+  /**
+   * Creates root record for a given record
+   *
+   * @param parentRow row containing id of the record
+   * @param data      dataset with all unique records in tree
+   * @param config    {@link HierarchyToRelationalConfig}
+   */
+  private void mapSelfToTreeRoot(Row parentRow, Dataset<Row> data, HierarchyToRelationalConfig config) {
+    String additionalFieldsForRoot = String.format("0 as %s,'%s' as %s,'%s' as %s ",
+                                                   config.levelField, config.falseValueField, config.topField,
+                                                   config.falseValueField, config.bottomField);
+    String parentProductField = "ParentProduct";
+    String childProductField = "ChildProduct";
+    String otherFields = "Supplier, Sales";
+    String selfAsParentQuery = String.format("select %s as %s, %s, %s as %s, %s, " +
+                                               "%s, %s from %s where %s=%s",
+                                             config.childField, config.parentField, config.childField,
+                                             childProductField, parentProductField, childProductField,
+                                             otherFields,
+                                             additionalFieldsForRoot, SQLUtils.NAME_HIERARCHY_TABLE, config.childField,
+                                             parentRow.get(0));
+    Dataset<Row> selfAsParentDf = data.sqlContext().sql(selfAsParentQuery).limit(1);
+    selfAsParentDf.registerTempTable(String.format("vt_%s_root_seed", parentRow.getInt(0)));
+  }
+
+  public JavaRDD<StructuredRecord> transform(SparkExecutionPluginContext sparkExecutionPluginContext,
+                                             JavaRDD<StructuredRecord> javaRDD) throws Exception {
+    Schema inputSchema = sparkExecutionPluginContext.getInputSchema();
+    if (outputSchema == null) {
+      outputSchema = config.generateOutputSchema(inputSchema);
```

Comment:
outputSchema will not be cached because a new instance of HierarchyToRelationalUtils is created every time transform() is called in https://github.com/data-integrations/hierarchical-to-relational/pull/1/files#diff-21e754e99391f007cb693985188f9450395e063c94e630b3c6f3694ce543a0bbR96